File: android/src/main/java/org/ligi/passandroid/model/comparator/PassTemporalDistanceComparator.java
package org.ligi.passandroid.model.comparator;

import org.ligi.passandroid.model.pass.Pass;
import org.threeten.bp.Duration;
import org.threeten.bp.LocalDateTime;

public class PassTemporalDistanceComparator extends PassByTimeComparator {

    @Override
    public int compare(Pass lhs, Pass rhs) {
        final int nullCheck = calculateCompareForNullValues(lhs, rhs);
        if (nullCheck == NO_NULL_VALUE) {

            final Duration durationLeft = Duration.between(LocalDateTime.now(), lhs.getCalendarTimespan().getFrom()).abs();
            final Duration durationRight = Duration.between(LocalDateTime.now(), rhs.getCalendarTimespan().getFrom()).abs();

            return durationLeft.compareTo(durationRight);
        } else {
            return nullCheck;
        }
    }
}

File: android/src/main/java/org/ligi/passandroid/model/comparator/PassTemporalDistanceComparator.kt
package org.ligi.passandroid.model.comparator

import org.ligi.passandroid.model.pass.Pass
import org.threeten.bp.Duration
import org.threeten.bp.LocalDateTime
import org.threeten.bp.ZonedDateTime

class PassTemporalDistanceComparator : PassByTimeComparator() {

    override fun compare(lhs: Pass, rhs: Pass): Int {
        return calculateCompareForNullValues(lhs, rhs, { leftDate: ZonedDateTime, rightDate: ZonedDateTime ->
            val durationLeft = Duration.between(LocalDateTime.now(), leftDate).abs()
            val durationRight = Duration.between(LocalDateTime.now(), rightDate).abs()

            return@calculateCompareForNullValues durationLeft.compareTo(durationRight)
        })
    }
}
